C# bindings for views

rekhoff opened the pull request:
# Description of Changes

Updates C# bindings to allow module authors to define `View`s and `AnonymousView`s using the pattern:
```csharp
[SpacetimeDB.View]
public static ExampleData? GetExampleDataById(ViewContext ctx, uint id)
{
    return ctx.Db.ExampleData.Id.Find(id);
}
	
[SpacetimeDB.View]
public static ExampleData? GetAnonymousExampleDataById(AnonymousViewContext ctx, uint id)
{
    return ctx.Db.ExampleData.Id.Find(id);
}
```
During publishing of a C# module, the views data will be converted into `RawViewDefV9`

# API and ABI breaking changes

No known breaking changes

# Expected complexity level and risk

2

# Testing

- [X] Locally tested locally adding the above sample pattern to the `sdks\csharp\examples~\regression-tests\server` test and performing a `dotnet build` and a `spacetime publish test`, both of which succeeded.

rekhoff committed:
Renamed CSharp Codegen TableViews to TableAccessors to give space to new Views

rekhoff committed:
C3 View Bindings Pass 1

rekhoff committed:
Updated with additional changes and reworks

rekhoff committed:
RegisterView fixed, removed ViewStubs as separate files, suppressed `CS0436` warning

rekhoff committed:
Cleaned up unneeded includes and code paths

rekhoff committed:
Corrected RegisterView to use existing TypeRegistrar

rekhoff committed:
Adding new ViewDispatchers during registration

rekhoff committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Fixes for module-test-cs to match the Rust view work

JasonAtClockwork committed:
Several fixes for CI and small bug on visibility for readonly accessors

JasonAtClockwork committed:
Generate missing updates on ClientApi, Quickstart, and Regression

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Update ensure_same_schema test to skip UpdateView for C#

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

Signed-off-by: Jason Larabie <jason@clockworklabs.io>

joshua-spacetime commented on sdks/csharp/examples~/regression-tests/server/Lib.cs:
```diff
@@ -15,6 +15,18 @@ public partial struct ExampleData
         public uint Indexed;
     }
 
+    [SpacetimeDB.View]
```
Note: Views must have an explicit name.

Also, as of right now, views must be made explicitly public in rust, although I have been considering just making public the default.

JasonAtClockwork committed:
Add error descriptors for View name requirement and public forced to true

JasonAtClockwork committed:
Update module test cs to properly label the View

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Add handling of View Indexes

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

joshua-spacetime commented:
@JasonAtClockwork could you also disable parameters for now? We can re-enable once we have sql support for them.

JasonAtClockwork committed:
Remove View arguments beyond the ViewContext

JasonAtClockwork committed:
Merge branch 'rekhoff/csharp-view-bindings' of https://github.com/clockworklabs/SpacetimeDB into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Removed IsAnonymous from the [View] attribute and left it to inferring from the context parameter

joshua-spacetime reviewed (commented):
@JasonAtClockwork Can you try to copy over some of the negative compile tests that are in `bindings/tests/ui/views.rs` into `bindings-csharp/Codegen.Tests/fixtures/diag/Lib.cs`? You don't have to copy all of them over. I mainly just want to test that compilation fails if we try to access the wrong methods on `ViewContext` and `AnonymousViewContext`, or return the wrong type.

JasonAtClockwork committed:
Removed GetIdentity() from IAnonymousViewContext, added ErrorDescriptor for improper View return types, and added more diag tests to mirror Rust tests

JasonAtClockwork commented:
> @JasonAtClockwork Can you try to copy over some of the negative compile tests that are in `bindings/tests/ui/views.rs` into `bindings-csharp/Codegen.Tests/fixtures/diag/Lib.cs`? You don't have to copy all of them over. I mainly just want to test that compilation fails if we try to access the wrong methods on `ViewContext` and `AnonymousViewContext`, or return the wrong type.

Copied most of the tests and fixed an issue with missing return type checks in the codegen

joshua-spacetime reviewed (approved):
I reviewed the tests and the runtime, but I did not review the codegen. I feel confident approving based on the test coverage, although I will outline a few more tests that would be useful to add before merging:

1. Can't obtain an identity from an AnonymousViewContext
2. Can't call `Iter()` on a table handle obtained from a read only context
3. Can't call mutations like delete on an index handle obtained from a read only context

JasonAtClockwork committed:
Removed Delete() from ReadOnly Indexes (no other mutations were previously included), removed Iter() from ReadOnly table accessors
Added tests for AnonymousContext GetIdentity(), ReadOnly Table Iter(), and ReadOnly Index Delete()

joshua-spacetime reviewed (approved):
Looks great! Thanks @JasonAtClockwork

JasonAtClockwork commented on crates/bindings-csharp/Codegen.Tests/fixtures/diag/Lib.cs:
```diff
@@ -1,4 +1,5 @@
 using System.ComponentModel;
+using System.Security.Cryptography.X509Certificates;
```
No.. my VS Code seems to add this "for" me I'll fix that up and confirm other files

JasonAtClockwork committed:
Removed improper using

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

JasonAtClockwork committed:
Merge branch 'master' into rekhoff/csharp-view-bindings

Signed-off-by: Jason Larabie <jason@clockworklabs.io>

cloutiertyler commented:
All tests are passing. I'm force merging this one.
